### ISSUE TITLE
other boolean operations?

### ISSUE THREAD

**Issue by maximilianh** [kimwalisch/libpopcnt#2]
Sorry, this is not an issue actually, just a naive question: popcnt usually goes together with other boolean operations. I am wondering what your opinion on these is: shall I just run the operation as a normal loop over the whole array, one byte a time (because they're already quite fast with -O3). Or do you recommend a library with (AVX-?) optimized versions of AND/OR/XOR etc on the array?

BTW, with g++ 4.4.7 from 2010, the benchmark file just crashes with "illegal instruction". Not a problem, I just had to update my g++.

thanks!
Max

**Comment by maximilianh** [comment]
Oh and my concrete use case here is set intersection, so I'm looking for a simple AND followed by your popcnt.

**Comment by kimwalisch** [comment]
> BTW, with g++ 4.4.7 from 2010, the benchmark file just crashes with "illegal instruction". Not a problem, I just had to update my g++.

Thanks for the bug report! I will try to fix the bug in the next days...

> shall I just run the as a normal loop over the whole array, one byte a time (because they're already quite fast). Or do you recommend a library with (AVX-?) optimized versions of AND/OR/XOR etc on the array?

It depends if this code is worth optimising to you (normally you start optimising a function or a hot loop if it is a bottleneck of your application). I am not sure if there is any library out there that fits your specific use case so you may have to implement an optimized popcnt algorithm with AND/OR/XOR yourself. Instead of doing one byte a time you could do 8-bytes a time using e.g. ```uin64_t```. ```libpopcnt.h``` offers 2 (undocumented) popcnt functions for 64-bit words which may be useful to you:

```C++
/* hardware POPCNT instruction */
uint64_t popcnt64(uint64_t x);

/* portable integer only algorithm */
uint64_t popcount64(uint64_t x);
```

Using the first function you could most likely code a simple much faster popcnt algorithm with AND than your current one byte a time.

**Comment by maximilianh** [comment]
Thanks! Hm...at first I didn't understand your reply, let me try to rephrase this to see if I understood:

I was thinking that I would simply AND  two large bitarrays followed by one POPCNT on the result. What you suggest instead is doing this piece-by-piece, right? Yes, I see why operating on long bitarrays - when one doesn't need it - is stupid... caching is bad and I waste bandwith. I can simply take each 512bit-part of the array, AND it and do POPCNT right away, on each 512-bit part.

Again, many thanks, I clearly didn't really think this through. I was so focussed on the POPCNT operation which I thought would always be the bottleneck, that I forgot that I don't need to operate on the full bitvector so your library may not be needed.

**Comment by kimwalisch** [comment]
>  I can simply take each 512bit-part of the array, AND it and do POPCNT right away, on each 512-bit part.

Yes, except that I suggested processing 64-bits (8-bytes) at a time using e.g. ```__builtin_popountll(x)``` (with compiler flag ```-mpopcnt```). Personally I would implement both solutions and benchmark them and then pick the fastest.

**Comment by maximilianh** [comment]
(I meant 256 bits, not 512, sorry). 

OK I think I understand:  with "processing 64bits" you mean a relatively normal C implementation with a normal integer "AND" operation followed by __builtin_popountll(x) (and keeping the sum of that while looping over the array).

The alternative would be: load 256 bits into the AVX2 registers, then AVX2 "AND" and AVX popcnt. I'm reading between the lines that it's not certain that AVX2 would be a lot faster than the simple 64bits approach, which is why you suggest trying the simple approach first.

Many thanks for your help! Performance optimization is a different world.

**Comment by kimwalisch** [comment]
> The alternative would be: load 256 bits into the AVX2 registers, then AVX2 "AND" and AVX popcnt. 

It will be very difficult to implement your algorithm using AVX2 because there is no vector popcount instruction available in AVX2. And you should not mix AVX code with the integer POPCNT instruction as this will give you poor performance (in general you should not mix vector instructions with non vector instructions). Also AVX code is very tricky, you need to take care of memory alignment, not all CPUs support AVX/AVX2 and so on.

Hence I suggest you code a simple algorithm using  ```__builtin_popountll(x)```, it will be very fast and you will be able to code it in a short amount of time.

**Comment by kimwalisch** [comment]
> BTW, with g++ 4.4.7 from 2010, the benchmark file just crashes with "illegal instruction". Not a problem, I just had to update my g++.

I just compiled and tested libpopcnt using g++ 4.4.7 and it works without any issues! Now I should also test on a PC without POPCNT instruction to make sure the CPUID detection code works correctly...

```sh
$ CC=gcc-4.4.7 CXX=g++-4.4.7 cmake ..
$ make test
Running tests...
Test project /home/kim/Desktop/libpopcnt-master/build
    Start 1: test1
1/2 Test #1: test1 ............................   Passed   12.94 sec
    Start 2: test2
2/2 Test #2: test2 ............................   Passed    1.86 sec

100% tests passed, 0 tests failed out of 2

Total Test time (real) =  14.80 sec

$./benchmark 
Status: 100%
Seconds: 11.1614

**Comment by maximilianh** [comment]
I tried it on Centos6.8, gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17), but
who is using these old gcc version anymore? I wouldn't worry.

**Comment by kimwalisch** [comment]
Thanks!

**Comment by kimwalisch** [comment]
I was able to reproduce the error:

```sh
$ g++ -v
gcc version 4.4.7 20120313 (Red Hat 4.4.7-18) (GCC)

$ cmake .
$ make
$ ./benchmark 
Illegal instruction
```

The error is caused by the ```xgetbv``` instruction that is used to detect if the CPU supports AVX2:

```C++
static inline int check_xcr0_ymm()
{
  int xcr0;
  int mask = XSTATE_SSE | XSTATE_YMM;
#if defined(_MSC_VER)
  xcr0 = (int) _xgetbv(0);
#else
  __asm__ ("xgetbv" : "=a" (xcr0) : "c" (0) : "%edx" );
#endif
  return (xcr0 & mask) == mask;
}
```

The fix is to only compile this function if the compiler supports AVX2:

```C++
#if defined(HAVE_AVX2)

static inline int check_xcr0_ymm()
{
  // ...
}

#endif
```

**Comment by kimwalisch** [comment]
Fixed https://github.com/kimwalisch/libpopcnt/commit/172c187b8cc9ff37227c30ed40da21d5ee63cd98
